### ISSUE TITLE
Add `SourceText.From(StringBuilder)` API

### ISSUE THREAD

**Issue by Neme12** [dotnet/roslyn#61326]
## Background and Motivation

This is an API proposal for #49082. I was asked by @333fred to create a separate issue for the API. PR is here: #61251
Pinging @chsienki 

Source generators often build up the source code using `StringBuilder`. There's currently no overload of `AddSource` or `SourceText.From` that takes a `StringBuilder`, so users either have to use `ToString()`, which can be problematic for very large strings or provide their own `TextReader` implementation over a `StringBuilder` to do this correctly, which isn't trivial or obvious.

## Proposed API

```diff
namespace Microsoft.CodeAnalysis.Text
{
    public abstract class SourceText
    {
        // New method:
+       public static SourceText From(StringBuilder stringBuilder, Encoding? encoding = null, SourceHashAlgorithm checksumAlgorithm = SourceHashAlgorithm.Sha1);

        // Existing methods for comparison:
        public static SourceText From(string text, Encoding? encoding = null, SourceHashAlgorithm checksumAlgorithm = SourceHashAlgorithm.Sha1);
        public static SourceText From(TextReader reader, int length, Encoding? encoding = null, SourceHashAlgorithm checksumAlgorithm = SourceHashAlgorithm.Sha1);
        public static SourceText From(Stream stream, Encoding? encoding = null, SourceHashAlgorithm checksumAlgorithm = SourceHashAlgorithm.Sha1, bool throwIfBinaryDetected = false, bool canBeEmbedded = false);
        public static SourceText From(byte[] buffer, int length, Encoding? encoding = null, SourceHashAlgorithm checksumAlgorithm = SourceHashAlgorithm.Sha1, bool throwIfBinaryDetected = false, bool canBeEmbedded = false);
     }
}
```

## Usage Examples

``` C#
var builder = new StringBuilder();
builder.AppendLine("using System;");
builder.AppendLine("Console.WriteLine(\"Hello, World!\");");
var sourceText = SourceText.From(builder);
```
Note that this would *not* create a `StringBuilderText` or store the `StringBuilder` in the `SourceText`. To ensure the immutability of `SourceText`, this would copy the text from the `StringBuilder` into the internal representation for `SourceText` - either `StringText` or `LargeText`, depending on the size, like the other non-string overloads do. Calling this overload would be equivalent to calling `SourceText From(TextReader reader, int length)` with a custom `TextReader` implementation that reads from the `StringBuilder`. The purpose of this API would be to let the user avoid having to allocate an enormous string (if the text is really large) by using `SourceText.From(stringBuilder.ToString())`.

## Questions
In addition to this, should we also add a corresponding `AddSource` overload like this?
```diff
namespace Microsoft.CodeAnalysis
{
    public readonly struct GeneratorExecutionContext
    {
        // New method:
+       public void AddSource(string hintName, StringBuilder source);

        // Existing methods for comparison:
        public void AddSource(string hintName, string source);
        public void AddSource(string hintName, SourceText sourceText);
    }
}
```
This method could be a shortcut to `SourceText.From(StringBuilder)`, just like the existing overload that takes a `string`. This isn't necessary, but it could help discoverability when writing source generators (which is the main scenario for this). I assume that if users only see overloads of `AddSource` that take a `string` and `SourceText`, rather than exploring how to create a `SourceText`, most users would likely call `ToString()` on the `StringBuilder`, which is a problematic way to do this for really large texts. If there wasn't an overload that takes `string` today, this wouldn't be an issue, as users would have to find out how to create a `SourceText` anyway and would then see the `StringBuilder` overload of `SourceText.From` next to the `string` one, but that's not the case, so this method might be really useful to aid in discoverability and help the user do the right thing.

## Alternative Designs

As opposed to this proposal, we could instead improve `SourceTextWriter` and make it public, or create a new type - `SourceTextBuilder`, and encourage users to build up the generated source code that way as opposed to using a `StringBuilder`. That way, we could make building up the `SourceText` a little more efficient for really large texts (many times larger than `SourceText.LargeObjectHeapLimitInChars` - 40 * 1024 `char`s).

I actually benchmarked this with Roslyn's `CSharpSyntaxGenerator`. This generator generates source files that are about 3.6 MB in size in total. The methods that actually generate the texts use a `TextWriter`, so it was easy to swap one for the other. I compared the current implementation that uses a `StringTextWriter` to get a `StringBuilder` from which `SourceText` is then created, to instead using `LargeTextWriter`. Here are the results:

```
|    Method |                  Job |              Runtime |     Mean |    Error |   StdDev |
|---------- |--------------------- |--------------------- |---------:|---------:|---------:|
|  Generate |             .NET 6.0 |             .NET 6.0 | 33.13 ms | 0.642 ms | 0.687 ms |
| Generate2 |             .NET 6.0 |             .NET 6.0 | 29.37 ms | 0.319 ms | 0.299 ms |
|  Generate |        .NET Core 3.1 |        .NET Core 3.1 | 48.81 ms | 0.949 ms | 1.559 ms |
| Generate2 |        .NET Core 3.1 |        .NET Core 3.1 | 40.07 ms | 0.644 ms | 0.571 ms |
|  Generate | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 57.10 ms | 1.140 ms | 1.171 ms |
| Generate2 | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 47.72 ms | 0.917 ms | 1.126 ms |
```

`Generate` is the original implementation that uses a `StringBuilder`, whereas `Generate2` uses `LargeTextWriter`. As we can see, using `LargeTextWriter` can indeed have some perf benefits.

There's an issue though - `SourceTextWriter` currently has 2 completely different implementations based on the length of the text, which is usually unknown in scenarios like generating code. This could be changed so that the type could dynamically switch strategy - it could write to an expanding array, and if more than 40 * 1024 characters are written, the implementation could switch so that it would start creating the chunks for `LargeText`.

But while we can make this a little more efficient when creating really large texts, the concern here is that this would be really hard to make as efficient as the `StringBuilder` approach for smaller to medium texts. That's because `StringBuilder` uses some internal methods to efficiently create a string and copy the data into it. I tried implementing a custom `StringBuilder`-like type for appending `string`s and comparing it to the built-in implementation, and the only way I could get at least close to the performance of `StringBuilder` was to use the same approach as `StringBuilder` does and create new chunks when adding more data than the available capacity, and that was on .NET (Core) only using the new `string.Create` API. Switching from an implementation like this to the one needed for large texts would require copying all of the data written so far into a different structure (into a single large chunk) at a certain threshold, which would have a negative perf impact in and of itself.

The only way to make this as efficient as `StringBuilder` for medium or small texts would be to always use `LargeText`, even for smaller texts, and when appending more data, to keep adding new chunks with a certain maximum size rather than copying the data into a fixed-sized chunk at a certain threshold - so `LargeText` would need to be able to store chunks that aren't of equal sizes. This might be a good win, and could actually be even faster than using `StringBuilder` and then creating a `string`, because the data could be kept as-is within the chunks rather than copying it into a `string`, except for a copy as a result of trimming the last chunk. Of course if we used `LargeText` even for smaller texts, that might then slightly decrease the performance of reading from the `SourceText`, so it's not clear if this would actually be a win.

Also, `StringBuilder`, as a built-in type, especially on .NET (Core), is really optimized and has a really efficient implementation for appending primitive types by using `ISpanFormattable` , it also has support for a custom interpolated string handler in .NET 6, etc. The type is close to being as efficient as can be and even more investments could go into it, not only with perf but new features as well. With a custom `SourceTextWriter`, it would be hard to get that level of support for writing all sorts of types just as efficiently, and would require specialized implementations and methods that are .NET (Core) only, such as support for writing `ReadOnlySpan<char>` and `ISpanFormattable`, in order for these operations to not be slower than when using `StringBuilder`.

## Risks

I can't think of any.

**Comment by Neme12** [comment]
Actually maybe adding a `SourceTextWriter` is a better solution after all. Why add an inferior API when we could add one that could be more efficient - sure, it will only be more efficient for really large texts, but the most likely scenario where people will have really large files is source generated ones. Although this API could still be useful when people are already using a `StringBuilder`.

**Comment by chsienki** [comment]
# API Review

- This is useful as its avoids a potentially single large allocation
- Still has copies however
  - Would it be more useful to just directly build up the source text 
  - Feels like a BCL hole, but we can plug it somewhat
    - If we expose a builder type ourselves then if/when the BCL hole is filled we still have to maintain a parallel impl
    - If we only expose `StringBuilder` and keep the reader internal then we could just replace the internal call with the BCL impl
  - What is the urgency for this? 
    - If we don't ship it as part of 17.3 (unlikely at this point) then consumers won't be able to rely on it for another few versions anyway
    - That gives us time to evaluate the `SourceTextWriter` like approach and potentially make an even better experience for authors than just `StringBuilder`


Consensus: Let's wait on this, explore what the `SourceTextWriter` alternative looks like and revisit once we have an API shape and perf data for it.

**Comment by jnm2** [comment]
We have support from @tannergooding to bring `StringBuilder.MoveChunksToStringBuilder()` to API review. This would allow Roslyn to create an API such as:

```cs
public static SourceText DrainFrom(StringBuilder stringBuilder)
{
    return new StringBuilderText(stringBuilder.MoveChunksToStringBuilder());
}
```
([StringBuilderText already exists](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/Text/StringBuilderText.cs))

The name of the Roslyn method will need to reflect that the user's builder is getting cleared.

The benefits of lighting up such an API are:
- There is less churn for users to adopt the performance win.
- An performance analyzer warning on `SourceText.From(stringBuilder.ToString())` is more doable, by the same token, because it can come with an easy automatic fix (switch to `SourceText.DrainFrom(stringBuilder)`)
- TextWriter-based APIs are much more limited than StringBuilder APIs. StringBuilder has AppendJoin, and lets you examine already-generated text, and allows mutation of already-generated text. StringBuilder also makes it super easy to chain a bunch of appends together.

**Comment by jnm2** [comment]
Of course, to expose this to users, Roslyn would need an additional `net11.0` target framework build and would need `#if NET11_0_OR_GREATER` around the new API.

**Comment by tannergooding** [comment]
We've approved the `StringBuilder` API as `StringBuilder.MoveChunks(source)` which produces a new StringBuilder containing the chunks of `source`. `source` is then cleareded (no longer tracks the original chunks)

**Comment by Neme12** [comment]
Why do we need to move the chunks instead of using the original chunks the StringBuilder already has?

**Comment by tannergooding** [comment]
Because you don't know how many users have an alias to the original `StringBuilder` and therefore could mutate the chunks.

You have to take the data and move it as an O(1) operation under a new string builder object to guarantee the right ownership semantics exist and therefore no one can mutate it but you (until you hand it out to other people).

**Comment by Neme12** [comment]
If only we had an [OwnershipTransfer] attribute and analyzers so that the endless defensive copying wasn't needed 😄

**Comment by tannergooding** [comment]
You still need "move" like APIs, such as what StringBuilder exposed, in such a world. It's also not trivially possible to make happen in existing code as its a breaking change.

This is also notably not really "copying", that is if you have a `StringBuilder` that has a length of say `9001` characters, this doesn't copy every individual character. It's simply "moving" the data from the old aliased builder to the new unaliased builder, so its essentially 0 cost other than the allocation for the string builder type itself.

i.e. we aren't defining `StringBuilder&&`, because that's too problematic, instead we've effectively defined `class StringBuilder { T&& chunks; }` instead; which keeps things nice non-breaking (or rather no changes to the existing `StringBuilder` contract).

**Comment by Neme12** [comment]
I understand that this is moving & not copying. I was just saying that it's sad that we need such an API in the first place.  An analyzer could see that the parameter is marked with [OwnershipTransfer], and all it would mean is that the caller cannot use their variable anymore, it now belongs to the callee. No move APIs necessary - just don't touch the variable anymore, it's not yours anymore.

**Comment by tannergooding** [comment]
Such an attribute becomes transitive and isn’t usable in a great deal of contexts, particularly with existing code. An analyzer would not solve the problem

**Comment by Neme12** [comment]
I don't understand why not and why it would be transitive. I have such an attribute in my projects - just without the analyzer, to mark when I'm passing a disposable to a class or function to say it's that function's responsibility to dispose of it now, not the caller's, or when returning a disposable from a function to say that it's now the caller's responsibility to dispose of it. But you are smart, so I have to believe you 😄 And we don't have to go on about this.

**Comment by tannergooding** [comment]
Because you have to ensure that no one else has access to the data. Simply this becomes illegal
```csharp
M(builder);
N(builder);

void M(StringBuilder builder) { ... }
void N([OwnershipTransfer] StringBuilder builder) { ... }
```

This is because `M` can capture `builder` anywhere and therefore an alias to `builder` can exist that they don't know becomes illegal to use after `N` is called.

References in .NET default to `shared` and the concept of "ownership" makes them `unique` instead. If you cannot guarantee that uniqueness, then you cannot transfer ownership; and in order to guarantee that uniqueness you rather have to transitively propagate that information up.

This is also why `ref structs` and `ref T` parameters/returns have so many special rules and attributes, because there is default assumptions as to what can capture or not capture things; which change based on whether the method takes or returns other byrefs. You then need `scoped` and `[UnscopedRef]` to indicate that the lifetime of the data is and whether something else can capture/alias the data or not, and therefore whether its memory safe.

**Comment by Neme12** [comment]
@tannergooding Thanks, I didn't think of that. Nevertheless, I will keep my attribute at least as a marker, especially for disposables.

**Comment by CyrusNajmabadi** [comment]
There's also the core issue around cooperative (and best effort) approaches to safety, versus enforced, and impossible-to-bypass-outside-of-unsafe approaches.  

I strongly like the idea of attributes and linters that you can use to best-effort provide the semantics you want, within the boundaries of your domain, while recognizing it can be violated.  But for core types like this, and for the sanity of roslyn maintainers, we want actual true enforcement that cannot be safely bypassed.

**Comment by Neme12** [comment]
@CyrusNajmabadi Yup. Internally, even Roslyn has the [NonCopyable] and [NonDefaultable] attributes. They are still useful, even if the enforcement isn't perfect. At the very least as markers so that users of the type know.
